--- a/kernel/ifftw.h
+++ b/kernel/ifftw.h
@@ -174,8 +174,8 @@
 #    define STACK_MALLOC(T, p, n)				\
      {								\
          p = (T)alloca((n) + MIN_ALIGNMENT);			\
-         p = (T)(((uintptr_t)p + (MIN_ALIGNMENT - 1)) &	\
-               (~(uintptr_t)(MIN_ALIGNMENT - 1)));		\
+         p = (T)((char *)p + ((-(uintptr_t)p) &		\
+               (uintptr_t)(MIN_ALIGNMENT - 1)));		\
      }
 #    define STACK_FREE(n) 
 #  else /* HAVE_ALLOCA && !defined(MIN_ALIGNMENT) */
@@ -1031,7 +1031,7 @@
 R *X(taint)(R *p, INT s);
 R *X(join_taint)(R *p1, R *p2);
 #define TAINT(p, s) X(taint)(p, s)
-#define UNTAINT(p) ((R *) (((uintptr_t) (p)) & ~(uintptr_t)3))
+#define UNTAINT(p) ((R *) ((char *)(p) - (((uintptr_t)(p)) & 3)))
 #define TAINTOF(p) (((uintptr_t)(p)) & 3)
 #define JOIN_TAINT(p1, p2) X(join_taint)(p1, p2)
 #else
--- a/kernel/kalloc.c
+++ b/kernel/kalloc.c
@@ -58,7 +58,7 @@
 {
      void *p0, *p;
      if (!(p0 = malloc(n + MIN_ALIGNMENT))) return (void *) 0;
-     p = (void *) (((uintptr_t) p0 + MIN_ALIGNMENT) & (~((uintptr_t) (MIN_ALIGNMENT - 1))));
+     p = (void *) ((char *)p0 + MIN_ALIGNMENT - ((uintptr_t)p0 & (MIN_ALIGNMENT - 1)));
      *((void **) p - 1) = p0;
      return p;
 }
--- a/simd-support/taint.c
+++ b/simd-support/taint.c
@@ -27,9 +27,9 @@
 R *X(taint)(R *p, INT s)
 {
      if (((unsigned)s * sizeof(R)) % ALIGNMENT)
-	  p = (R *) (PTRINT(p) | TAINT_BIT);
+	  p = (R *) ((char *)p + (TAINT_BIT & ~PTRINT(p)));
      if (((unsigned)s * sizeof(R)) % ALIGNMENTA)
-	  p = (R *) (PTRINT(p) | TAINT_BITA);
+	  p = (R *) ((char *)p + (TAINT_BITA & ~PTRINT(p)));
      return p;
 }
 
@@ -38,6 +38,6 @@
 R *X(join_taint)(R *p1, R *p2)
 {
      A(UNTAINT(p1) == UNTAINT(p2));
-     return (R *)(PTRINT(p1) | PTRINT(p2));
+     return (R *)((char *)p1 + (TAINTOF(p2) & ~TAINTOF(p1)));
 }
 #endif
