diff --git a/tools/meson.build b/tools/meson.build
--- a/tools/meson.build
+++ b/tools/meson.build
@@ -12,7 +12,7 @@
 if cc.get_id() == 'msvc'
   python_cmd = '/usr/bin/env ' + python.path()
 else
-  python_cmd = '/usr/bin/env python@0@'.format(python.language_version().split('.')[0])
+  python_cmd = python.path()
 endif
 
 tool_output = []
diff --git a/meson.build b/meson.build
--- a/meson.build
+++ b/meson.build
@@ -248,7 +248,7 @@
 
 # The tests will also run, which is not possible if they
 # were built for a different architecture.
-if not meson.is_cross_build()
+if meson.can_run_host_binaries()
   subdir('tests')
 endif
 
