diff --git a/src/config/shlib.conf b/src/config/shlib.conf
index 75b7cc3af63d..2765e2057a0d 100644
--- a/src/config/shlib.conf
+++ b/src/config/shlib.conf
@@ -427,7 +427,7 @@ mips-*-netbsd*)
 	LDCOMBINE='$(CC) -shared -fPIC -Wl,-h,$(LIBPREFIX)$(LIBBASE)$(SHLIBSEXT) $(UNDEF_CHECK)'
 	UNDEF_CHECK='-Wl,--no-undefined'
 	# $(EXPORT_CHECK) runs export-check.pl when in maintainer mode.
-	LDCOMBINE_TAIL='-Wl,--version-script binutils.versions $(EXPORT_CHECK)'
+	LDCOMBINE_TAIL='--version-script=binutils.versions $(EXPORT_CHECK)'
 	SHLIB_EXPORT_FILE_DEP=binutils.versions
 	RPATH_FLAG='-Wl,--enable-new-dtags -Wl,-rpath -Wl,'
 	# For cases where we do have dependencies on other libraries
diff --git a/src/util/verto/ev.c b/src/util/verto/ev.c
index 87f90670fc12..ec9ec35a5a82 100644
--- a/src/util/verto/ev.c
+++ b/src/util/verto/ev.c
@@ -179,6 +179,8 @@
 
 #include <signal.h>
 
+#include <stdfil.h>
+
 #ifdef EV_H
 # include EV_H
 #else
@@ -653,6 +655,11 @@ struct signalfd_siginfo
   #include <intrin.h> /* fence functions _ReadBarrier, also bit search functions _BitScanReverse */
 #endif
 
+/* Let's play it safe with fences. There's no point in being fancy. */
+#define ECB_MEMORY_FENCE zfence()
+#define ECB_MEMORY_FENCE_ACQUIRE zfence()
+#define ECB_MEMORY_FENCE_RELEASE zfence()
+
 #ifndef ECB_MEMORY_FENCE
   #if ECB_GCC_VERSION(2,5) || defined __INTEL_COMPILER || (__llvm__ && __GNUC__) || __SUNPRO_C >= 0x5110 || __SUNPRO_CC >= 0x5110
     #if __i386 || __i386__
