diff --git a/src/mod_cgi.c b/src/mod_cgi.c
index 0c9c575..3e6623b 100644
--- a/src/mod_cgi.c
+++ b/src/mod_cgi.c
@@ -1,5 +1,9 @@
 #include "first.h"
 
+#ifdef __FILC__
+#include <stdfil.h>
+#endif
+
 #include "base.h"
 #include "stat_cache.h"
 #include "http_kv.h"
@@ -929,8 +933,13 @@ static int cgi_create_env(request_st * const r, plugin_data * const p, handler_c
 		envp = (char **)env->offsets;
 		envp[env->oused] = NULL;
 		const uintptr_t baseptr = (uintptr_t)env->b->ptr;
-		for (i = 0; i < env->oused; ++i)
+		for (i = 0; i < env->oused; ++i) {
+		  #ifdef __FILC__
+			envp[i] = zmkptr(env->b->ptr, (uintptr_t)envp[i] + baseptr);
+		  #else
 			envp[i] += baseptr;
+		  #endif
+		}
 
 		/* set up args */
 		i = 0;
diff --git a/src/server.c b/src/server.c
index a9977f2..8d88b4d 100644
--- a/src/server.c
+++ b/src/server.c
@@ -1902,8 +1902,9 @@ static int server_main_setup (server * const srv, int argc, char **argv) {
 	/* clear Linux ambient capabilities, if any had been granted
 	 * (avoid leaking privileges to CGI or other subprocesses) */
 	if (prctl(PR_CAP_AMBIENT, PR_CAP_AMBIENT_CLEAR_ALL, 0L, 0L, 0L) < 0
-           /* not supported before linux 4.3 / on some emulators (e.g. Cloud Run 1st gen) */
-           && errno != EINVAL) {
+           /* not supported before linux 4.3 / on some emulators / Fil-C */
+           && errno != EINVAL
+           && errno != ENOSYS) {
 		log_perror(srv->errh, __FILE__, __LINE__,
 		  "prctl(PR_CAP_AMBIENT, PR_CAP_AMBIENT_CLEAR_ALL)");
 		return -1;
