diff --git a/lib/sigsegv.c b/lib/sigsegv.c
index d0519814ebf2..4ac0377868e5 100644
--- a/lib/sigsegv.c
+++ b/lib/sigsegv.c
@@ -1417,7 +1417,7 @@ int
 stackoverflow_install_handler (stackoverflow_handler_t handler,
                                void *extra_stack, size_t extra_stack_size)
 {
-#if HAVE_STACK_OVERFLOW_RECOVERY
+#if HAVE_STACK_OVERFLOW_RECOVERY && !defined(__FILC__)
 # if HAVE_STACKVMA
   if (!stack_top)
     {
