diff --git a/girepository/gdump.c b/girepository/gdump.c
index 363575cf13ea..1dccfa66d3a3 100644
--- a/girepository/gdump.c
+++ b/girepository/gdump.c
@@ -171,7 +171,7 @@ invoke_error_quark (GModule *self, const char *symbol, GError **error)
 		   G_FILE_ERROR,
 		   G_FILE_ERROR_FAILED,
 		   "Failed to find symbol '%s'", symbol);
-      return G_TYPE_INVALID;
+      return (GQuark)G_TYPE_INVALID;
     }
 
   return sym ();
@@ -220,23 +220,23 @@ value_to_string (const GValue *value)
     {
       GType value_type = G_VALUE_TYPE (value);
 
-      switch (G_TYPE_FUNDAMENTAL (value_type))
+      switch ((uintptr_t)G_TYPE_FUNDAMENTAL (value_type))
         {
-        case G_TYPE_BOXED:
+        case (uintptr_t)G_TYPE_BOXED:
           if (g_value_get_boxed (value) == NULL)
             return NULL;
           else
             return value_transform_to_string (value);
           break;
 
-        case G_TYPE_OBJECT:
+        case (uintptr_t)G_TYPE_OBJECT:
           if (g_value_get_object (value) == NULL)
             return NULL;
           else
             return value_transform_to_string (value);
           break;
 
-        case G_TYPE_POINTER:
+        case (uintptr_t)G_TYPE_POINTER:
           return NULL;
 
         default:
@@ -551,24 +551,24 @@ dump_type (GType type, const char *symbol, FILE *out)
    */
   GObjectClass *gobject_class = g_type_class_ref (G_TYPE_OBJECT);
 
-  switch (g_type_fundamental (type))
+  switch ((uintptr_t)g_type_fundamental (type))
     {
-    case G_TYPE_OBJECT:
+    case (uintptr_t)G_TYPE_OBJECT:
       dump_object_type (type, symbol, out);
       break;
-    case G_TYPE_INTERFACE:
+    case (uintptr_t)G_TYPE_INTERFACE:
       dump_interface_type (type, symbol, out);
       break;
-    case G_TYPE_BOXED:
+    case (uintptr_t)G_TYPE_BOXED:
       dump_boxed_type (type, symbol, out);
       break;
-    case G_TYPE_FLAGS:
+    case (uintptr_t)G_TYPE_FLAGS:
       dump_flags_type (type, symbol, out);
       break;
-    case G_TYPE_ENUM:
+    case (uintptr_t)G_TYPE_ENUM:
       dump_enum_type (type, symbol, out);
       break;
-    case G_TYPE_POINTER:
+    case (uintptr_t)G_TYPE_POINTER:
       /* GValue, etc.  Just skip them. */
       break;
     default:
diff --git a/girepository/ginvoke.c b/girepository/ginvoke.c
index 5d423b1e2066..28d7d956ac36 100644
--- a/girepository/ginvoke.c
+++ b/girepository/ginvoke.c
@@ -43,48 +43,48 @@ value_to_ffi_type (const GValue *gvalue, gpointer *value)
   GType type = g_type_fundamental (G_VALUE_TYPE (gvalue));
   g_assert (type != G_TYPE_INVALID);
 
-  switch (type)
+  switch ((uintptr_t)type)
     {
-    case G_TYPE_BOOLEAN:
-    case G_TYPE_CHAR:
-    case G_TYPE_INT:
+    case (uintptr_t)G_TYPE_BOOLEAN:
+    case (uintptr_t)G_TYPE_CHAR:
+    case (uintptr_t)G_TYPE_INT:
       rettype = &ffi_type_sint;
       *value = (gpointer)&(gvalue->data[0].v_int);
       break;
-    case G_TYPE_UCHAR:
-    case G_TYPE_UINT:
+    case (uintptr_t)G_TYPE_UCHAR:
+    case (uintptr_t)G_TYPE_UINT:
       rettype = &ffi_type_uint;
       *value = (gpointer)&(gvalue->data[0].v_uint);
       break;
-    case G_TYPE_STRING:
-    case G_TYPE_OBJECT:
-    case G_TYPE_BOXED:
-    case G_TYPE_POINTER:
-    case G_TYPE_PARAM:
+    case (uintptr_t)G_TYPE_STRING:
+    case (uintptr_t)G_TYPE_OBJECT:
+    case (uintptr_t)G_TYPE_BOXED:
+    case (uintptr_t)G_TYPE_POINTER:
+    case (uintptr_t)G_TYPE_PARAM:
       rettype = &ffi_type_pointer;
       *value = (gpointer)&(gvalue->data[0].v_pointer);
       break;
-    case G_TYPE_FLOAT:
+    case (uintptr_t)G_TYPE_FLOAT:
       rettype = &ffi_type_float;
       *value = (gpointer)&(gvalue->data[0].v_float);
       break;
-    case G_TYPE_DOUBLE:
+    case (uintptr_t)G_TYPE_DOUBLE:
       rettype = &ffi_type_double;
       *value = (gpointer)&(gvalue->data[0].v_double);
       break;
-    case G_TYPE_LONG:
+    case (uintptr_t)G_TYPE_LONG:
       rettype = &ffi_type_slong;
       *value = (gpointer)&(gvalue->data[0].v_long);
       break;
-    case G_TYPE_ULONG:
+    case (uintptr_t)G_TYPE_ULONG:
       rettype = &ffi_type_ulong;
       *value = (gpointer)&(gvalue->data[0].v_ulong);
       break;
-    case G_TYPE_INT64:
+    case (uintptr_t)G_TYPE_INT64:
       rettype = &ffi_type_sint64;
       *value = (gpointer)&(gvalue->data[0].v_int64);
       break;
-    case G_TYPE_UINT64:
+    case (uintptr_t)G_TYPE_UINT64:
       rettype = &ffi_type_uint64;
       *value = (gpointer)&(gvalue->data[0].v_uint64);
       break;
@@ -116,46 +116,46 @@ g_value_to_ffi_return_type (const GValue *gvalue,
 
   *value = (gpointer)&(ffi_value->v_long);
 
-  switch (type) {
-  case G_TYPE_CHAR:
+  switch ((uintptr_t)type) {
+  case (uintptr_t)G_TYPE_CHAR:
     rettype = &ffi_type_sint8;
     break;
-  case G_TYPE_UCHAR:
+  case (uintptr_t)G_TYPE_UCHAR:
     rettype = &ffi_type_uint8;
     break;
-  case G_TYPE_BOOLEAN:
-  case G_TYPE_INT:
+  case (uintptr_t)G_TYPE_BOOLEAN:
+  case (uintptr_t)G_TYPE_INT:
     rettype = &ffi_type_sint;
     break;
-  case G_TYPE_UINT:
+  case (uintptr_t)G_TYPE_UINT:
     rettype = &ffi_type_uint;
     break;
-  case G_TYPE_STRING:
-  case G_TYPE_OBJECT:
-  case G_TYPE_BOXED:
-  case G_TYPE_POINTER:
-  case G_TYPE_PARAM:
+  case (uintptr_t)G_TYPE_STRING:
+  case (uintptr_t)G_TYPE_OBJECT:
+  case (uintptr_t)G_TYPE_BOXED:
+  case (uintptr_t)G_TYPE_POINTER:
+  case (uintptr_t)G_TYPE_PARAM:
     rettype = &ffi_type_pointer;
     break;
-  case G_TYPE_FLOAT:
+  case (uintptr_t)G_TYPE_FLOAT:
     rettype = &ffi_type_float;
     *value = (gpointer)&(ffi_value->v_float);
     break;
-  case G_TYPE_DOUBLE:
+  case (uintptr_t)G_TYPE_DOUBLE:
     rettype = &ffi_type_double;
     *value = (gpointer)&(ffi_value->v_double);
     break;
-  case G_TYPE_LONG:
+  case (uintptr_t)G_TYPE_LONG:
     rettype = &ffi_type_slong;
     break;
-  case G_TYPE_ULONG:
+  case (uintptr_t)G_TYPE_ULONG:
     rettype = &ffi_type_ulong;
     break;
-  case G_TYPE_INT64:
+  case (uintptr_t)G_TYPE_INT64:
     rettype = &ffi_type_sint64;
     *value = (gpointer)&(ffi_value->v_int64);
     break;
-  case G_TYPE_UINT64:
+  case (uintptr_t)G_TYPE_UINT64:
     rettype = &ffi_type_uint64;
     *value = (gpointer)&(ffi_value->v_uint64);
     break;
@@ -179,50 +179,50 @@ static void
 g_value_from_ffi_value (GValue           *gvalue,
                         const GIArgument *value)
 {
-  switch (g_type_fundamental (G_VALUE_TYPE (gvalue))) {
-  case G_TYPE_INT:
+  switch ((uintptr_t)g_type_fundamental (G_VALUE_TYPE (gvalue))) {
+  case (uintptr_t)G_TYPE_INT:
       g_value_set_int (gvalue, (gint)value->v_long);
       break;
-  case G_TYPE_FLOAT:
+  case (uintptr_t)G_TYPE_FLOAT:
       g_value_set_float (gvalue, (gfloat)value->v_float);
       break;
-  case G_TYPE_DOUBLE:
+  case (uintptr_t)G_TYPE_DOUBLE:
       g_value_set_double (gvalue, (gdouble)value->v_double);
       break;
-  case G_TYPE_BOOLEAN:
+  case (uintptr_t)G_TYPE_BOOLEAN:
       g_value_set_boolean (gvalue, (gboolean)value->v_long);
       break;
-  case G_TYPE_STRING:
+  case (uintptr_t)G_TYPE_STRING:
       g_value_set_string (gvalue, (gchar*)value->v_pointer);
       break;
-  case G_TYPE_CHAR:
+  case (uintptr_t)G_TYPE_CHAR:
       g_value_set_schar (gvalue, (gchar)value->v_long);
       break;
-  case G_TYPE_UCHAR:
+  case (uintptr_t)G_TYPE_UCHAR:
       g_value_set_uchar (gvalue, (guchar)value->v_ulong);
       break;
-  case G_TYPE_UINT:
+  case (uintptr_t)G_TYPE_UINT:
       g_value_set_uint (gvalue, (guint)value->v_ulong);
       break;
-  case G_TYPE_POINTER:
+  case (uintptr_t)G_TYPE_POINTER:
       g_value_set_pointer (gvalue, (gpointer)value->v_pointer);
       break;
-  case G_TYPE_LONG:
+  case (uintptr_t)G_TYPE_LONG:
       g_value_set_long (gvalue, (glong)value->v_long);
       break;
-  case G_TYPE_ULONG:
+  case (uintptr_t)G_TYPE_ULONG:
       g_value_set_ulong (gvalue, (gulong)value->v_ulong);
       break;
-  case G_TYPE_INT64:
+  case (uintptr_t)G_TYPE_INT64:
       g_value_set_int64 (gvalue, (gint64)value->v_int64);
       break;
-  case G_TYPE_UINT64:
+  case (uintptr_t)G_TYPE_UINT64:
       g_value_set_uint64 (gvalue, (guint64)value->v_uint64);
       break;
-  case G_TYPE_BOXED:
+  case (uintptr_t)G_TYPE_BOXED:
       g_value_set_boxed (gvalue, (gpointer)value->v_pointer);
       break;
-  case G_TYPE_PARAM:
+  case (uintptr_t)G_TYPE_PARAM:
       g_value_set_param (gvalue, (gpointer)value->v_pointer);
       break;
   default:
diff --git a/tests/gimarshallingtests.c b/tests/gimarshallingtests.c
index c50988f1fbf8..7410c1617ab7 100644
--- a/tests/gimarshallingtests.c
+++ b/tests/gimarshallingtests.c
@@ -6387,7 +6387,7 @@ void
 gi_marshalling_tests_param_spec_in_bool (const GParamSpec *param)
 {
   g_assert (G_IS_PARAM_SPEC (param));
-  g_assert_cmpint (G_PARAM_SPEC_VALUE_TYPE (param), ==, G_TYPE_BOOLEAN);
+  g_assert_cmpint ((uintptr_t)G_PARAM_SPEC_VALUE_TYPE (param), ==, (uintptr_t)G_TYPE_BOOLEAN);
   g_assert_cmpstr (g_param_spec_get_name ((GParamSpec *) param), ==, "mybool");
 }
 
diff --git a/tests/scanner/regress.c b/tests/scanner/regress.c
index c137bccf152f..de222ae8c129 100644
--- a/tests/scanner/regress.c
+++ b/tests/scanner/regress.c
@@ -11,6 +11,8 @@
 
 #include "regress.h"
 
+#include <stdfil.h>
+
 static gboolean abort_on_error = TRUE;
 
 #define ASSERT_VALUE(condition)  \
@@ -2349,7 +2351,7 @@ regress_test_obj_class_init (RegressTestObjClass *klass)
                    0     /* n_params */,
                    NULL  /* param_types */);
 
-  param_types[0] = regress_test_simple_boxed_a_get_gtype() | G_SIGNAL_TYPE_STATIC_SCOPE;
+  param_types[0] = zorptr(regress_test_simple_boxed_a_get_gtype(), (uintptr_t)G_SIGNAL_TYPE_STATIC_SCOPE);
   klass->test_signal_with_static_scope_arg =
     g_signal_newv ("test-with-static-scope-arg",
                    G_TYPE_FROM_CLASS (gobject_class),
@@ -2653,7 +2655,7 @@ regress_test_obj_class_init (RegressTestObjClass *klass)
       g_signal_new ("sig-with-gerror", G_TYPE_FROM_CLASS (klass),
                     G_SIGNAL_RUN_LAST, 0, NULL, NULL,
                     g_cclosure_marshal_generic, G_TYPE_NONE, 1,
-                    G_TYPE_ERROR | G_SIGNAL_TYPE_STATIC_SCOPE);
+                    (GType)((uintptr_t)G_TYPE_ERROR | (uintptr_t)G_SIGNAL_TYPE_STATIC_SCOPE));
 
   gobject_class->set_property = regress_test_obj_set_property;
   gobject_class->get_property = regress_test_obj_get_property;
