diff --git a/src/hb-gobject-enums.cc.tmpl b/src/hb-gobject-enums.cc.tmpl
index 87a11dd407eb..cff55d0019ad 100644
--- a/src/hb-gobject-enums.cc.tmpl
+++ b/src/hb-gobject-enums.cc.tmpl
@@ -55,9 +55,9 @@
 GType
 @enum_name@_get_type ()
 {
-  static gsize type_id = 0;
+  static gpointer type_id = 0;
 
-  if (g_once_init_enter (&type_id))
+  if (g_once_init_enter_pointer (&type_id))
     {
       static const G@Type@Value values[] = {
 /*** END value-header ***/
@@ -71,10 +71,10 @@ GType
       };
       GType id =
         g_@type@_register_static (g_intern_static_string ("@EnumName@"), values);
-      g_once_init_leave (&type_id, id);
+      g_once_init_leave_pointer (&type_id, id);
     }
 
-  return type_id;
+  return (GType) type_id;
 }
 
 /*** END value-tail ***/
diff --git a/src/hb-gobject-structs.cc b/src/hb-gobject-structs.cc
index d66de0b237f4..6c3f2d500313 100644
--- a/src/hb-gobject-structs.cc
+++ b/src/hb-gobject-structs.cc
@@ -64,14 +64,14 @@
 GType \
 hb_gobject_##name##_get_type () \
 { \
-   static gsize type_id = 0; \
-   if (g_once_init_enter (&type_id)) { \
+   static gpointer type_id = 0; \
+   if (g_once_init_enter_pointer (&type_id)) { \
       GType id = g_boxed_type_register_static (g_intern_static_string ("hb_" #name "_t"), \
 					       (GBoxedCopyFunc) copy_func, \
 					       (GBoxedFreeFunc) free_func); \
-      g_once_init_leave (&type_id, id); \
+      g_once_init_leave_pointer (&type_id, id); \
    } \
-   return type_id; \
+   return (GType) type_id; \
 }
 
 #define HB_DEFINE_OBJECT_TYPE(name) \
