diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5ad071832722..b4d5f051871e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -512,14 +512,14 @@ endif()
 if(UNIX AND NOT APPLE)
   file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/conftest.map "VERS_1 { global: *; };")
   set(CMAKE_REQUIRED_FLAGS
-    "-Wl,--version-script,${CMAKE_CURRENT_BINARY_DIR}/conftest.map")
+    "--version-script=${CMAKE_CURRENT_BINARY_DIR}/conftest.map")
   check_c_source_compiles("int main(void) { return 0; }" HAVE_VERSION_SCRIPT)
   set(CMAKE_REQUIRED_FLAGS)
   file(REMOVE ${CMAKE_CURRENT_BINARY_DIR}/conftest.map)
   if(HAVE_VERSION_SCRIPT)
     message(STATUS "Linker supports GNU-style version scripts")
-    set(MAPFLAG "-Wl,--version-script,")
-    set(TJMAPFLAG "-Wl,--version-script,")
+    set(MAPFLAG "--version-script=")
+    set(TJMAPFLAG "--version-script=")
   else()
     message(STATUS "Linker does not support GNU-style version scripts")
     if(CMAKE_SYSTEM_NAME STREQUAL "SunOS")
