diff --git a/catalog.c b/catalog.c
index 96d3e16..6a6b342 100644
--- a/catalog.c
+++ b/catalog.c
@@ -110,7 +110,7 @@ typedef xmlCatalogEntry *xmlCatalogEntryPtr;
 struct _xmlCatalogEntry {
     struct _xmlCatalogEntry *next;
     struct _xmlCatalogEntry *parent;
-    struct _xmlCatalogEntry *children;
+    struct _xmlCatalogEntry *_Atomic children;
     xmlCatalogEntryType type;
     xmlChar *name;
     xmlChar *value;
