diff --git a/libpam/Makefile.am b/libpam/Makefile.am
index 505900b5fe0d..cde2bc65a263 100644
--- a/libpam/Makefile.am
+++ b/libpam/Makefile.am
@@ -5,7 +5,7 @@
 AM_CFLAGS = -DDEFAULT_MODULE_PATH=\"$(SECUREDIR)/\" -DLIBPAM_COMPILE \
 	-DLOCALEDIR=\"$(localedir)\" -I$(top_srcdir)/libpam_internal/include \
 	-I$(srcdir)/include $(LIBPRELUDE_CFLAGS) $(ECONF_CFLAGS) \
-	-DPAM_VERSION=\"$(VERSION)\" -DSYSCONFDIR=\"$(sysconfdir)\" \
+	-DPAM_VERSION=\"$(VERSION)\" -DSYSCONFDIR=\"$(sysconfdir)\" -DLIBDIR=\"$(libdir)\" \
 	$(WARN_CFLAGS)
 
 CLEANFILES = *~
diff --git a/libpam/pam_private.h b/libpam/pam_private.h
index 77dc5a69878c..143f4aeb858a 100644
--- a/libpam/pam_private.h
+++ b/libpam/pam_private.h
@@ -25,11 +25,11 @@
 
 /* the Linux-PAM configuration file */
 
-#define PAM_CONFIG         "/etc/pam.conf"
-#define PAM_CONFIG_D       "/etc/pam.d"
-#define PAM_CONFIG_DF      "/etc/pam.d/%s"
-#define PAM_CONFIG_DIST_D  "/usr/lib/pam.d"
-#define PAM_CONFIG_DIST_DF "/usr/lib/pam.d/%s"
+#define PAM_CONFIG         SYSCONFDIR"/pam.conf"
+#define PAM_CONFIG_D       SYSCONFDIR"/pam.d"
+#define PAM_CONFIG_DF      SYSCONFDIR"/pam.d/%s"
+#define PAM_CONFIG_DIST_D  LIBDIR"/pam.d"
+#define PAM_CONFIG_DIST_DF LIBDIR"/pam.d/%s"
 #ifdef VENDORDIR
 #define PAM_CONFIG_DIST2_D  VENDORDIR"/pam.d"
 #define PAM_CONFIG_DIST2_DF VENDORDIR"/pam.d/%s"
