diff --git a/tools/encoding/Makefile b/tools/encoding/Makefile
new file mode 100644
index 000000000000..361239e9c23a
--- /dev/null
+++ b/tools/encoding/Makefile
@@ -0,0 +1,110 @@
+#
+# This file is a Makefile to compile all the encoding files.
+#
+# Run "make" to compile all the encoding files (*.txt,*.esc) into the
+# format that Tcl can use (*.enc).  It is your responsibility to move the
+# encoding files to the appropriate place ($TCL_ROOT/library/encoding
+#
+# The .txt files in this directory come from the Unicode CD and are covered
+# by the following copyright notice:
+#
+#---------------------------------------------------------------------------
+#
+# Copyright (c) 1996 Unicode, Inc.  All Rights reserved.
+#
+# This file is provided as-is by Unicode, Inc. (The Unicode Consortium).
+# No claims are made as to fitness for any particular purpose.  No
+# warranties of any kind are expressed or implied.  The recipient
+# agrees to determine applicability of information provided.  If this
+# file has been provided on magnetic media by Unicode, Inc., the sole
+# remedy for any claim will be exchange of defective media within 90
+# days of receipt.
+#
+# Recipient is granted the right to make copies in any form for
+# internal distribution and to freely use the information supplied
+# in the creation of products supporting Unicode.  Unicode, Inc.
+# specifically excludes the right to re-distribute this file directly
+# to third parties or other organizations whether for profit or not.
+#
+# In other words:  Don't put this file on the Internet.  People who want to
+# get it over the Internet should do so directly from ftp://unicode.org.  They
+# can therefore be assured of getting the most recent and accurate version.
+#
+#----------------------------------------------------------------------------
+#
+# The txt2enc program built by this makefile is used to compile individual
+# .txt files into .enc files, the format that Tcl understands for encoding
+# files.  This compilation to a different format is allowed by the above
+# restriction.
+#
+# The files shiftjis.txt and jis0208.txt were modified from the original
+# ones provided on the Unicode CD.  The double-width backslash character
+# 0x815F in these two Japanese encodings was being mapped to Unicode 005C
+# (REVERSE SOLIDUS), the normal backslash character.  They have been
+# changed to map 0x815F to Unicode FF3C (FULLWIDTH REVERSE SOLIDUS) and let
+# the regular backslash character map to itself.  This follows how cp932
+# behaves.
+#
+# Copyright (c) 1998 Sun Microsystems, Inc.
+#
+# See the file "license.terms" for information on usage and redistribution
+# of this file, and for a DISCLAIMER OF ALL WARRANTIES.
+#
+# SCCS: @(#) Makefile 1.1 98/01/28 11:41:36
+#
+
+EUC_ENCODINGS = euc-cn.txt euc-kr.txt euc-jp.txt
+
+encodings: clean txt2enc $(EUC_ENCODINGS)
+	@echo Compiling encoding files.
+	@for p in *.esc; do \
+	    base=`echo $$p | sed 's/\..*$$//'`; \
+	    echo $$base.enc; \
+	    echo "# Encoding file: $$base, escape-driven" > $$base.enc; \
+	    echo "E" >> $$base.enc; \
+	    cat $$p >> $$base.enc; \
+	done
+	@for p in *.txt; do \
+	    enc=`echo $$p | sed 's/\..*$$/\.enc/'`; \
+	    echo $$enc; \
+	    ./txt2enc -e 0 -u 1 $$p > $$enc; \
+	done
+	@echo
+	@echo Compiling special versions of encoding files.
+	@for p in ascii.txt; do \
+	    enc=`echo $$p | sed 's/\..*$$/\.enc/'`; \
+	    echo $$enc; \
+	    ./txt2enc -m $$p > $$enc; \
+	done
+	@for p in jis0208.txt; do \
+	    enc=`echo $$p | sed 's/\..*$$/\.enc/'`; \
+	    echo $$enc; \
+	    ./txt2enc -e 1 -u 2 $$p > $$enc; \
+	done
+	@for p in symbol.txt dingbats.txt macDingbats.txt; do \
+	    enc=`echo $$p | sed 's/\..*$$/\.enc/'`; \
+	    echo $$enc; \
+	    ./txt2enc -e 0 -u 1 -s $$p > $$enc; \
+	done
+
+clean:
+	@rm -f txt2enc *.enc $(EUC_ENCODINGS)
+
+txt2enc: txt2enc.c
+	@gcc -o txt2enc txt2enc.c
+
+euc-jp.txt: ascii.txt jis0208.txt Makefile
+	@echo Building euc-jp.txt from components.
+	@cat ascii.txt > euc-jp.txt
+	@grep '^0x[A-F]' jis0201.txt | sed 's/0x/0x8E/' >> euc-jp.txt
+	@cat jis0208.txt | awk 'BEGIN {print "ibase=16"} index($$1,"#") != 1 {print substr($$2,3) "+" 8080; print substr($$3,3)}' | bc | awk '{ str = $$1; getline; printf "0x%04x 0x%04x\n", str, $$0}' >> euc-jp.txt
+
+euc-kr.txt: ascii.txt ksc5601.txt Makefile
+	@echo Building euc-kr.txt from components.
+	@cat ascii.txt > euc-kr.txt
+	@cat ksc5601.txt | awk 'BEGIN {print "ibase=16"} index($$1,"#") != 1 {print substr($$1,3) "+" 8080; print substr($$2,3)}' | bc | awk '{ str = $$1; getline; printf "0x%04x 0x%04x\n", str, $$0}' >> euc-kr.txt
+
+euc-cn.txt: ascii.txt gb2312.txt Makefile
+	@echo Building euc-cn.txt from components.
+	@cat ascii.txt > euc-cn.txt
+	@cat gb2312.txt | awk 'BEGIN {print "ibase=16"} index($$1,"#") != 1 {print substr($$1,3) "+" 8080; print substr($$2,3)}' | bc | awk '{ str = $$1; getline; printf "0x%04x 0x%04x\n", str, $$0}' >> euc-cn.txt
diff --git a/win/coffbase.txt b/win/coffbase.txt
new file mode 100644
index 000000000000..96d03e1fe15d
--- /dev/null
+++ b/win/coffbase.txt
@@ -0,0 +1,43 @@
+;
+; This file defines the virtual base addresses for the Dynamic Link Libraries
+; that are part of the Tcl system.  The first token on a line is the key (or name
+; of the DLL) and the second token is the virtual base address, in hexadecimal.
+; The third token is the maximum size of the DLL image file, including symbols.
+;
+; Using a specified "preferred load address" should speed loading time by avoiding
+; relocations (NT supported only).  It is assumed extension authors will contribute
+; their modules to this grand-master list.  You can use the dumpbin utility with
+; the /headers option to get the "size of image" data (already in hex).  If the
+; maximum size is too small a linker warning will occur.  Modules can overlap when
+; they're mutually exclusive.  This info is placed in the DLL's PE header by the
+; linker with the `-base:@$(TCLDIR)\win\coffbase.txt,<key>` option.
+
+tcl		0x10000000	0x00200000
+tcldde		0x10200000	0x00010000
+tclreg		0x10210000	0x00010000
+tk		0x10220000	0x00200000
+expect		0x10480000	0x00080000
+itcl		0x10500000	0x00080000
+itk		0x10580000	0x00080000
+bltlite		0x10600000	0x00080000
+blt		0x10680000	0x00080000
+iocpsock	0x10700000	0x00080000
+tls		0x10780000	0x00100000
+winico		0x10880000	0x00010000
+sample		0x108B0000	0x00010000
+tile		0x10900000	0x00080000
+memchan		0x109D0000	0x00010000
+tdom		0x109E0000	0x00080000
+tclvfs		0x10A70000	0x00010000
+tkvideo		0x10B00000	0x00010000
+tclsdl		0x10B20000	0x00080000
+vqtcl		0x10C00000	0x00010000
+tdbc		0x10C40000	0x00010000
+thread		0x10C80000	0x00020000
+nsf		0x10ca0000	0x00080000
+;
+; insert new packages here
+;
+snack		0x1E000000	0x00400000
+sound		0x1E400000	0x00400000
+snackogg	0x1E800000	0x00200000
diff --git a/win/tclsh.exe.manifest.in b/win/tclsh.exe.manifest.in
new file mode 100644
index 000000000000..8b06fce7c1d5
--- /dev/null
+++ b/win/tclsh.exe.manifest.in
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0"
+	xmlns:asmv3="urn:schemas-microsoft-com:asm.v3">
+    <assemblyIdentity
+	    version="@TCL_WIN_VERSION@"
+	    processorArchitecture="@MACHINE@"
+	    name="Tcl.tclsh"
+	    type="win32"
+	/>
+    <description>Tcl command line shell (tclsh)</description>
+    <trustInfo xmlns="urn:schemas-microsoft-com:asm.v3">
+	<security>
+	    <requestedPrivileges>
+		<requestedExecutionLevel
+			level="asInvoker"
+			uiAccess="false"
+		    />
+	    </requestedPrivileges>
+	</security>
+    </trustInfo>
+    <compatibility xmlns="urn:schemas-microsoft-com:compatibility.v1">
+	<application>
+	    <!-- Windows 10 -->
+	    <supportedOS Id="{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}"/>
+	    <!-- Windows 8.1 -->
+	    <supportedOS Id="{1f676c76-80e1-4239-95bb-83d0f6d0da78}"/>
+	    <!-- Windows 8 -->
+	    <supportedOS Id="{4a2f28e3-53b9-4441-ba9c-d69d4a4a6e38}"/>
+	    <!-- Windows 7 -->
+	    <supportedOS Id="{35138b9a-5d96-4fbd-8e2d-a2440225f93a}"/>
+	    <!-- Windows Vista -->
+	    <supportedOS Id="{e2011457-1546-43c5-a5fe-008deee3d3f0}"/>
+	</application>
+    </compatibility>
+    <asmv3:application>
+	<asmv3:windowsSettings
+		xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">
+	    <dpiAware>true</dpiAware>
+	</asmv3:windowsSettings>
+    </asmv3:application>
+    <dependency>
+	<dependentAssembly>
+	    <assemblyIdentity
+		    type="win32"
+		    name="Microsoft.Windows.Common-Controls"
+		    version="6.0.0.0"
+		    processorArchitecture="@MACHINE@"
+		    publicKeyToken="6595b64144ccf1df"
+		    language="*"
+		/>
+	</dependentAssembly>
+    </dependency>
+</assembly>
diff --git a/win/tclsh.ico b/win/tclsh.ico
new file mode 100644
index 000000000000..e2543187962f
Binary files /dev/null and b/win/tclsh.ico differ
diff --git a/win/tclsh.rc b/win/tclsh.rc
new file mode 100644
index 000000000000..685bebd1af5d
--- /dev/null
+++ b/win/tclsh.rc
@@ -0,0 +1,81 @@
+//
+// Version Resource Script
+//
+
+#include <winver.h>
+#include <tcl.h>
+
+//
+// build-up the name suffix that defines the type of build this is.
+//
+#if TCL_THREADS
+#define SUFFIX_THREADS	    "t"
+#else
+#define SUFFIX_THREADS	    ""
+#endif
+
+#if STATIC_BUILD
+#define SUFFIX_STATIC	    "s"
+#else
+#define SUFFIX_STATIC	    ""
+#endif
+
+#if DEBUG && !UNCHECKED
+#define SUFFIX_DEBUG	    "g"
+#else
+#define SUFFIX_DEBUG	    ""
+#endif
+
+#define SUFFIX		    SUFFIX_THREADS SUFFIX_STATIC SUFFIX_DEBUG
+
+
+LANGUAGE 0x9, 0x1	/* LANG_ENGLISH, SUBLANG_DEFAULT */
+
+VS_VERSION_INFO VERSIONINFO
+ FILEVERSION 	TCL_MAJOR_VERSION,TCL_MINOR_VERSION,TCL_RELEASE_LEVEL,TCL_RELEASE_SERIAL
+ PRODUCTVERSION TCL_MAJOR_VERSION,TCL_MINOR_VERSION,TCL_RELEASE_LEVEL,TCL_RELEASE_SERIAL
+ FILEFLAGSMASK 	0x3fL
+#ifdef DEBUG
+ FILEFLAGS 	VS_FF_DEBUG
+#else
+ FILEFLAGS 	0x0L
+#endif
+ FILEOS 	VOS__WINDOWS32
+ FILETYPE 	VFT_APP
+ FILESUBTYPE 	0x0L
+BEGIN
+    BLOCK "StringFileInfo"
+    BEGIN
+        BLOCK "040904b0"
+        BEGIN
+            VALUE "FileDescription", "Tclsh Application\0"
+            VALUE "OriginalFilename", "tclsh" STRINGIFY(TCL_MAJOR_VERSION) STRINGIFY(TCL_MINOR_VERSION) SUFFIX ".exe\0"
+            VALUE "FileVersion", TCL_PATCH_LEVEL
+            VALUE "LegalCopyright", "Copyright \251 1987-2022 Regents of the University of California and other parties\0"
+            VALUE "ProductName", "Tcl " TCL_VERSION " for Windows\0"
+            VALUE "ProductVersion", TCL_PATCH_LEVEL
+        END
+    END
+    BLOCK "VarFileInfo"
+    BEGIN
+        VALUE "Translation", 0x409, 1200
+    END
+END
+
+//
+// Icon
+//
+
+tclsh                      ICON    DISCARDABLE     "tclsh.ico"
+
+//
+// This is needed for Windows 8.1 onwards.
+//
+
+#ifndef RT_MANIFEST
+#define RT_MANIFEST     24
+#endif
+#ifndef CREATEPROCESS_MANIFEST_RESOURCE_ID
+#define CREATEPROCESS_MANIFEST_RESOURCE_ID 1
+#endif
+CREATEPROCESS_MANIFEST_RESOURCE_ID RT_MANIFEST "tclsh.exe.manifest"
