diff --git a/Src/Zle/zle_keymap.c b/Src/Zle/zle_keymap.c
index 2389ab7542bf..f963f31bae05 100644
--- a/Src/Zle/zle_keymap.c
+++ b/Src/Zle/zle_keymap.c
@@ -85,6 +85,7 @@ struct keymap {
 struct key {
     HashNode next;	/* next in hash chain */
     char *nam;		/* key sequence (metafied) */
+    int flags;          /* field from hashnode */
     Thingy bind;	/* binding of this key sequence */
     char *str;		/* string for send-string (metafied) */
     int prefixct;	/* number of sequences for which this is a prefix */
diff --git a/Src/mem.c b/Src/mem.c
index 5951e57ed071..07834d0f66e8 100644
--- a/Src/mem.c
+++ b/Src/mem.c
@@ -576,6 +576,9 @@ zheapptr(void *p)
 mod_export void *
 zhalloc(size_t size)
 {
+    if ((1))
+        return malloc(size);
+    
     Heap h, hp = NULL;
     size_t n;
 #ifdef ZSH_VALGRIND
@@ -704,6 +707,12 @@ hrealloc(char *p, size_t old, size_t new)
 	return zhalloc(new);
 #endif
 
+    if ((1)) {
+        char* result = malloc(new);
+        memcpy(result, p, old < new ? old : new);
+        return result;
+    }
+    
     /* find the heap with p */
 
     queue_signals();
diff --git a/stamp-h.in b/stamp-h.in
new file mode 100644
index 000000000000..8b137891791f
--- /dev/null
+++ b/stamp-h.in
@@ -0,0 +1 @@
+
